gcp_variant_transforms/libs/variant_partition.py

```diff
@@ -34,26 +43,216 @@
+    if not self._interval_tree:
+      # interval tree has not been initiated.
+      self._interval_tree = intervaltree.IntervalTree()
+    else:
+      # Check the existing interval tree for possible overlap with new interval.
+      if self._interval_tree.overlaps_range(start, end):
+        raise ValueError('Cannot add overlapping region {}-{}'.
+                         format(start, end))
+    # If everything goes well we add the new region to the interval tree.
+    self._interval_tree.addi(start, end, index)
+
+  def add_index(self, index):
+    if self._interval_tree is not None:
+      raise ValueError('Can not add full chromosome to existing multi region.')
+    if self._index != -1:
+      raise ValueError('This instance has already been assinged an index.')
+    if index < 0:
+      raise ValueError('Index of a region cannot be negative {}'.format(index))
+    self._index = index
+
+  def get_index(self, pos=0):
+    if self._interval_tree is None and self._index == -1:
+      raise ValueError('This instance has not been assigned any index.')
+    if self._interval_tree is None:
+      return self._index
+    else:
+      matched_intervals = self._interval_tree.search(pos)
+      if len(matched_intervals) > 1:
+        raise ValueError('Search for pos {} returned more than 1 result'.
+                         format(pos))
+      if len(matched_intervals) == 1:
+        return next(iter(matched_intervals)).data
+      else:
+        return -1
+
+  @property
+  def is_full_chromosome(self):
+    return self._index != -1
+
+class VariantPartition(object):
+  """Partition variants based on their reference_name and/or position.
+
+  This class has 2 operating modes:
+    1) No config file is given (config_file_path == ''):
+        Automatically partition variants based on their reference_name.
+    2) A config file is given:
+        partitions variants based on input config file.
+  """
+
+  def __init__(self, config_file_path=''):
+    if not config_file_path:
+      if _DEFAULT_NUM_PARTITIONS <= _RESERVED_AUTO_PARTITIONS:
+        raise ValueError(
+            '_DEFAULT_NUM_PARTITIONS must be > _RESERVED_AUTO_PARTITIONS')
+      self._config_file_path_given = False
+      self._reference_name_to_partition_map = {}
+    else:
+      self._config_file_path_given = True
+      self._parse_config(config_file_path)
+
+  def _parse_config(self, config_file_path):
+    """Parses the given partition config file.
+
+    Args:
+      config_file_path: name of the input partition_config file.
+    Raises:
+      A ValueError if any of the expected config formats are violated.
+    """
+    def _parse_position(pos_str):
+      return int(pos_str.replace(',', ''))
+
+    with FileSystems.open(config_file_path, 'r') as f:
+      try:
+        partition_configs = yaml.load(f)
+      except yaml.YAMLError as e:
+        raise ValueError('Invalid yaml file: %s' % str(e))
+
+    if len(partition_configs) > _MAX_NUM_PARTITIONS:
```

Comment:
instead of adding ValueErrors throughout the actual logic, consider creating a validate() method that validates the content first. You can then go through the contents assuming it's valid. I think for some of them you still need to validate while parsing (e.g. checking for overlapping regions), but it would be simpler.

Reply:
done.
Note two of the ValueErrors are still remained in _parse_config, if we move those two to validate we will be basically implementing the whole parse logic in validate and only avoid recording the result. To make the distinction between the parse and validate, I left those two ValueErrors in parse. Please let me know what do you think.